lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -119,24 +119,33 @@ class _Visitor extends SimpleAstVisitor {
   }
 
   void check(AstNode node) {
+    bool checkStatements(AstNode child, NodeList<Statement> statements) {
+      var index = statements.indexOf(child as Statement);
+      for (var i = index - 1; i >= 0; i--) {
+        var s = statements.elementAt(i);
+        if (isMountedCheck(s)) {
+          return false;
+        } else if (isAsync(s)) {
+          rule.reportLint(node);
+          return true;
+        }
+      }
+      return true;
+    }
+
     // Walk back and look for an async gap that is not guarded by a mounted
     // property check.
     AstNode? child = node;
     while (child != null && child is! FunctionBody) {
       var parent = child.parent;
-      // todo (pq): refactor to handle SwitchCase's
       if (parent is Block) {
-        var statements = parent.statements;
-        var index = statements.indexOf(child as Statement);
-        for (var i = index - 1; i >= 0; i--) {
-          var s = statements.elementAt(i);
-          if (isMountedCheck(s)) {
-            return;
-          } else if (isAsync(s)) {
-            rule.reportLint(node);
-            return;
-          }
+        var keepChecking = checkStatements(child, parent.statements);
+        if (!keepChecking) {
+          return;
         }
+      } else if (parent is SwitchCase) {
+        checkStatements(child, parent.statements);
+        return;
```

Comment:
It seems wrong to not look above the switch statement. Is the rule behaving as expected for the following?
```
void awaitInSwitchCase_mountedCheckBeforeSwitch(BuildContext context) async {
    await Future<void>.delayed(Duration());
    if (!mounted) return;
    switch (1) {
      case 1:
        await Navigator.of(context).pushNamed('routeName'); // OK
        break;
    }
  }
```

Reply:
yep.  That works (and test case added) but I'm not surprised you're confused since this is getting unwieldy.  I'll take a refactoring conversation offline...

Reply:
(actually it worked but only by accident.)

I think it's actually fixed now but would still like to explore a general refactoring so will follow up on that.